core/lib/contract_verifier/src/lib.rs

```diff
@@ -556,7 +557,33 @@ impl ContractVerifier {
         request_id: usize,
         contract: &DeployedContractData,
         creation_bytecode: &[u8],
+        contract_identifier: &ContractIdentifier,
     ) -> Result<ConstructorArgs, ContractVerifierError> {
+        fn extract_arguments<'a>(
+            calldata: &'a [u8],
+            creation_bytecode: &'a [u8],
+            metadata_len: usize,
+        ) -> Result<&'a [u8], &'static str> {
+            if creation_bytecode.len() < metadata_len {
+                // This shouldn't normally happen, since we calculated contract identifier based on this code.
+                return Err("Creation bytecode doesn't fit metadata");
+            }
+            let creation_bytecode_without_metadata =
+                &creation_bytecode[..creation_bytecode.len() - metadata_len];
+
+            // Ensure equivalence of the creation bytecode (which can be different from the deployed bytecode).
+            // Note that metadata hash may still be different; this is checked by other part of the code.
+            let constructor_args_with_metadata = calldata
+                .strip_prefix(creation_bytecode_without_metadata)
+                .ok_or("Creation bytecode is different")?;
+
+            // Skip metadata to get to the constructor arguments.
+            if constructor_args_with_metadata.len() < metadata_len {
+                return Err("Provided bytecode has no metadata");
+            }
+            Ok(&constructor_args_with_metadata[metadata_len..])
```

Comment:
I'm not sure I understand how this works. If metadata can be different from one in `creation_bytecode`, what's the guarantee that it has the same length (if it's CBOR metadata)?

Reply:
The probability of lengths being different is negligibly low, IMHO (for that we need either different contract hash format or pre-release solc version -- I've never seen either of those in practice).

But sure, just in case we can protect the implementation against that.